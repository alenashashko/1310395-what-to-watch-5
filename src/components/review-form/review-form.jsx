import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import withRatingAndReviewText
  from '../../hocs/with-rating-and-review-text/with-rating-and-review-text';
import {getErrorText, getIsCommentLoadingStatus} from '../../store/selectors';
import {increaseBrightness} from '../../utils';

const TEXTAREA_BACKGROUND_COLOR_BRIGHTNESS_PERCENT_TO_INCREASE = 30;

const ReviewForm = (props) => {
  const {
    movieBackgroundColor,
    ratingValue,
    reviewText,
    onFormSubmit,
    onTextChange,
    onRatingChange,
    saveErrorText,
    isLoading
  } = props;
  const ratings = new Array(5).fill(null);

  const textareaColor = increaseBrightness(
      movieBackgroundColor,
      TEXTAREA_BACKGROUND_COLOR_BRIGHTNESS_PERCENT_TO_INCREASE
  );

  return (
    <form onSubmit={onFormSubmit} action="#" className="add-review__form">
      <div className="rating">
        <div className="rating__stars">
          {ratingValue === null ? (
            <input
              id={`star-hidden`}
              name="rating"
              className={`visually-hidden`}
              type="radio"
              defaultChecked={true}
            />
          ) : null}

          {ratings.map((it, index) => {
            const isChecked = ratingValue ? ratingValue === index + 1 : false;

            return (
              <React.Fragment key={index}>
                <input
                  id={`star-${index + 1}`}
                  className="rating__input"
                  type="radio"
                  value={index + 1}
                  name="rating"
                  checked={isChecked}
                  onChange={() => onRatingChange(index + 1)}
                />
                <label
                  className="rating__label" htmlFor={`star-${index + 1}`}>
                  {`Rating ${index + 1}`}
                </label>
              </React.Fragment>
            );
          })}
        </div>
      </div>

      <div
        className="add-review__text"
        style={{
          backgroundColor: `${textareaColor}`,
        }}
      >
        <textarea
          onChange={onTextChange}
          value={reviewText}
          className="add-review__textarea"
          name="review-text"
          id="review-text"
          placeholder="Review text"
          minLength="50"
          maxLength="400"
        >
        </textarea>
        <div className="add-review__submit">
          <button
            className="add-review__btn"
            type="submit"
            disabled={isLoading || ratingValue === null || reviewText.length < 50}
          >
            Post
          </button>
        </div>

      </div>
      {saveErrorText
        ? <div style={{
          color: `#ed8d8d`,
          textAlign: `center`,
          paddingTop: `20px`
        }}>{saveErrorText}</div> : null}
    </form>
  );
};

ReviewForm.propTypes = {
  movieBackgroundColor: PropTypes.string.isRequired,
  ratingValue: PropTypes.number,
  reviewText: PropTypes.string.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
  onTextChange: PropTypes.func.isRequired,
  onRatingChange: PropTypes.func.isRequired,
  saveErrorText: PropTypes.string,
  isLoading: PropTypes.bool
};

const mapStateToProps = (state) => ({
  saveErrorText: getErrorText(state),
  isLoading: getIsCommentLoadingStatus(state)
});

export {ReviewForm};
export default connect(mapStateToProps)(withRatingAndReviewText(ReviewForm));
