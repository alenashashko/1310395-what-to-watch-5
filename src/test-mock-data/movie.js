export const movie = {
  id: `1`,
  title: `Orlando`,
  picture: `/img/orlando.jpg`,
  backgroundPicture: `/img/orlando.jpg`,
  backgroundColor: `#e1b0b2`,
  poster: `/img/orlando.jpg`,
  genre: `Crime`,
  year: 2015,
  ratingScore: 7.8,
  ratingCount: 1000,
  src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  duration: 110,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Cras aliquet varius magna, non porta ligula feugiat eget.
  Fusce tristique felis at fermentum pharetra.
  Aliquam id orci ut lectus varius viverra.
  Nullam nunc ex, convallis sed finibus eget, sollicitudin eget ante.
  Phasellus eros mauris, condimentum sed nibh vitae, sodales efficitur ipsum.
  Sed blandit, eros vel aliquam faucibus, purus ex euismod diam, eu luctus nunc ante ut dui.
  Sed sed nisi sed augue convallis suscipit in sed felis.
  Aliquam erat volutpat.`,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`,
    `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`],
  isFavorite: false
};

export const movies = [{
  id: `1`,
  title: `Orlando`,
  picture: `/img/orlando.jpg`,
  backgroundPicture: `/img/orlando.jpg`,
  backgroundColor: `#e1b0b2`,
  poster: `/img/orlando.jpg`,
  genre: `Crime`,
  year: 2015,
  ratingScore: 7.8,
  ratingCount: 1000,
  src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  duration: 110,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Cras aliquet varius magna, non porta ligula feugiat eget.
  Fusce tristique felis at fermentum pharetra.
  Aliquam id orci ut lectus varius viverra.
  Nullam nunc ex, convallis sed finibus eget, sollicitudin eget ante.
  Phasellus eros mauris, condimentum sed nibh vitae, sodales efficitur ipsum.
  Sed blandit, eros vel aliquam faucibus, purus ex euismod diam, eu luctus nunc ante ut dui.
  Sed sed nisi sed augue convallis suscipit in sed felis.
  Aliquam erat volutpat.`,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`,
    `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`],
  isFavorite: false
},
{
  id: `2`,
  title: `Revenant`,
  picture: `/img/revenant.jpg`,
  backgroundPicture: `/img/revenant.jpg`,
  backgroundColor: `#bdad8f`,
  poster: `/img/revenant.jpg`,
  genre: `Romance`,
  year: 2010,
  ratingScore: 5.9,
  ratingCount: 850,
  src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  duration: 145,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Cras aliquet varius magna, non porta ligula feugiat eget.
  Fusce tristique felis at fermentum pharetra.
  Aliquam id orci ut lectus varius viverra.
  Nullam nunc ex, convallis sed finibus eget, sollicitudin eget ante.
  Phasellus eros mauris, condimentum sed nibh vitae, sodales efficitur ipsum.`,
  director: `Robert Zemeckis`,
  starring: [`Will Smith`, `Michael J. Fox`, `Harrison Ford`,
    `Sandra Bullock`, `Betty White`, `Tom Hanks`, `Denzel Washington`],
  isFavorite: false
}
];

export const genres = [`Comedies`, `Crime`, `Documentary`, `Dramas`, `Horror`,
  `Kids & Family`, `Romance`, `Sci-Fi`, `Thrillers`];

export const genre = `All genres`;
