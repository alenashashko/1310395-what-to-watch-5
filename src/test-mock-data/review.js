import {adaptCommentToClient} from '../services/adapters';

export const apiReviews = [
  {
    id: 1,
    user: {
      id: 11,
      name: `Jack`
    },
    rating: 2.4,
    comment: `It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie.`,
    date: `2020-11-03T13:38:44.769Z`
  },
  {
    id: 2,
    user: {
      id: 12,
      name: `Isaac`
    },
    rating: 7.2,
    comment: `A movie that will take you to another world full of emotions.`,
    date: `2020-10-23T13:38:44.769Z`
  }
];

export const review = adaptCommentToClient(apiReviews[0]);

export const reviews = apiReviews.map(adaptCommentToClient);

