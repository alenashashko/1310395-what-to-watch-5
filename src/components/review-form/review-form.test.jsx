import React from 'react';
import renderer from 'react-test-renderer';

import {ReviewForm} from './review-form';
import {review} from '../../test-mock-data/review';
import {movie} from '../../test-mock-data/movie';

const noop = () => {};

describe(`Should ReviewForm render correctly`, () => {
  it(`with save error text`, () => {
    const tree = renderer
      .create(
          <ReviewForm
            movieBackgroundColor={movie.backgroundColor}
            ratingValue={4}
            reviewText={review.text}
            onFormSubmit={noop}
            onTextChange={noop}
            onRatingChange={noop}
            saveErrorText="Lorem ipsum"
            isLoading={false}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`without save error text`, () => {
    const tree = renderer
      .create(
          <ReviewForm
            movieBackgroundColor={movie.backgroundColor}
            ratingValue={4}
            reviewText={review.text}
            onFormSubmit={noop}
            onTextChange={noop}
            onRatingChange={noop}
            saveErrorText={null}
            isLoading={false}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`with loading status`, () => {
    const tree = renderer
      .create(
          <ReviewForm
            movieBackgroundColor={movie.backgroundColor}
            ratingValue={4}
            reviewText={review.text}
            onFormSubmit={noop}
            onTextChange={noop}
            onRatingChange={noop}
            saveErrorText={null}
            isLoading={true}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
