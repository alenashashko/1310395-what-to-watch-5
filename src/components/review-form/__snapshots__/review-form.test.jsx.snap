// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewForm render correctly with loading status 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
    >
      <input
        checked={false}
        className="rating__input"
        id="star-1"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-2"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-3"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        checked={true}
        className="rating__input"
        id="star-4"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-5"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
    style={
      Object {
        "backgroundColor": "#9dcab8",
      }
    }
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie."
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={true}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;

exports[`Should ReviewForm render correctly with save error text 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
    >
      <input
        checked={false}
        className="rating__input"
        id="star-1"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-2"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-3"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        checked={true}
        className="rating__input"
        id="star-4"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-5"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
    style={
      Object {
        "backgroundColor": "#9dcab8",
      }
    }
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie."
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={false}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "color": "#ed8d8d",
        "paddingTop": "20px",
        "textAlign": "center",
      }
    }
  >
    Lorem ipsum
  </div>
</form>
`;

exports[`Should ReviewForm render correctly without save error text 1`] = `
<form
  action="#"
  className="add-review__form"
  onSubmit={[Function]}
>
  <div
    className="rating"
  >
    <div
      className="rating__stars"
    >
      <input
        checked={false}
        className="rating__input"
        id="star-1"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        className="rating__label"
        htmlFor="star-1"
      >
        Rating 1
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-2"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        className="rating__label"
        htmlFor="star-2"
      >
        Rating 2
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-3"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={3}
      />
      <label
        className="rating__label"
        htmlFor="star-3"
      >
        Rating 3
      </label>
      <input
        checked={true}
        className="rating__input"
        id="star-4"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={4}
      />
      <label
        className="rating__label"
        htmlFor="star-4"
      >
        Rating 4
      </label>
      <input
        checked={false}
        className="rating__input"
        id="star-5"
        name="rating"
        onChange={[Function]}
        type="radio"
        value={5}
      />
      <label
        className="rating__label"
        htmlFor="star-5"
      >
        Rating 5
      </label>
    </div>
  </div>
  <div
    className="add-review__text"
    style={
      Object {
        "backgroundColor": "#9dcab8",
      }
    }
  >
    <textarea
      className="add-review__textarea"
      id="review-text"
      maxLength="400"
      minLength="50"
      name="review-text"
      onChange={[Function]}
      placeholder="Review text"
      value="It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie."
    />
    <div
      className="add-review__submit"
    >
      <button
        className="add-review__btn"
        disabled={false}
        type="submit"
      >
        Post
      </button>
    </div>
  </div>
</form>
`;
